data/rulesets.ts

```diff
@@ -1604,6 +1604,79 @@ export const Rulesets: {[k: string]: FormatData} = {
 			pokemon.trapped = true;
 		},
 	},
+	crazyhouserule: {
+		effectType: 'Rule',
+		name: 'Crazyhouse Rule',
+		desc: "You can use the pokemon you directly KO.",
+		onValidateRule(value) {
+			const ruleTable = this.ruleTable;
+			const maxTeamSize = ruleTable.pickedTeamSize || ruleTable.maxTeamSize;
+			const numPlayers = (this.format.gameType === 'freeforall' || this.format.gameType === 'multi') ? 4 : 2;
+			const potentialMaxTeamSize = maxTeamSize * numPlayers;
+			if (potentialMaxTeamSize > 24) {
+				throw new Error(`Crazyhouse Rule cannot be added because a team can potentially have ${potentialMaxTeamSize} Pokemon on one team, which is more than the server limit of 24.`);
+			}
+		},
+		// In order to prevent a case of the clones, housekeeping is needed.
+		// This is especially needed to make sure one side doesn't end up with too many Pokemon.
+		onBeforeSwitchIn(pokemon) {
+			if (this.turn < 1 || !pokemon.side.faintedThisTurn) return;
+			pokemon.side.pokemon = pokemon.side.pokemon.filter(x => !(x.fainted && !x.m.outofplay));
+			for (let i = 0; i < pokemon.side.pokemon.length && i < 24; i++) {
+				pokemon.side.pokemon[i].position = i;
+			}
+		},
+		onFaint(target, source, effect) {
+			if (!target.m.numSwaps) {
+				target.m.numSwaps = 0;
+			}
+			target.m.numSwaps++;
+			if (effect && effect.effectType === 'Move' && source.side.pokemon.length < 24 &&
+                source.side !== target.side && target.m.numSwaps < 4) {
+				const Pokemon: typeof import('../sim/pokemon').Pokemon =
+                    require('../sim/pokemon').Pokemon;
+				const hpCost = this.clampIntRange(Math.floor((target.baseMaxhp * target.m.numSwaps) / 4), 1);
+				// Just in case(tm) and for Shedinja
+				if (hpCost === target.baseMaxhp) {
+					target.m.outofplay = true;
+					return;
+				}
+				source.side.pokemonLeft++;
+				source.side.pokemon.length++;
+
+				const newPoke = new Pokemon(target.set, source.side);
```

Comment:
Do you really need to use the Pokemon class here? can't you just deepClone the Pokemon?

Reply:
Calling deepClone causes a crash. This was the next best thing.

Reply:
Oh also, it does help immensely because it treats the new mon as as owned by the capturer, making adding it to the opponent's team easier to do without much hiccups.

Reply:
The old Gen 7 Super Staff Bros had code to swap Pokémon between sides without cloning them, but I don't know whether it's applicable to what you're trying to achieve here.

Reply:
> The old Gen 7 Super Staff Bros had code to swap Pokémon between sides without cloning them

can you point me to it? It could work but I couldn't find it after searching

Reply:
https://github.com/smogon/pokemon-showdown/blob/8768b2d7fb71840521f30e659d7fe717c3cd47f1/config/formats.js#L1941

Reply:
oh thanks! I can probably use that. The only problem might be pointing issues, but I am willing to give it a shot.

Reply:
> The old Gen 7 Super Staff Bros had code to swap Pokémon between sides without cloning them, but I don't know whether it's applicable to what you're trying to achieve here.
Trying to adapt the code into what I am trying to do causes an invalid switch error, which is why I have a clone to help prevent this sort of crash. It might be possible with more tinkering, but i can't think of any way off the top of my head.

Reply:
Looks like deepClone crashes because Pokemon objects reference a Side object, which are have a reference loop: https://github.com/smogon/pokemon-showdown/blob/master/sim/side.ts#L66

Unsure where to go from here for fixing this, or if we even should fix infinite recursion in deepClone.

Reply:
Deep cloning isn't meant to be used for class instances in the first place.

Reply:
We could get rid of the `Pokemon` dependency if we had am `addPokemon` function for it in `Side`.

Reply:
Why are you cloning in the first place? What do you want to carry over that isn't in the set?